Simplify and expand kinematic tests for bboxes

sfmig opened the pull request:
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
To ensure our existing kinematics methods play nicely with bboxes datasets.

**What does this PR do?**
- Simplifies kinematics tests
	- Refactored tests for displacement, velocity and acceleration into a single parametrised test for computing kinematic variables
	- Separated tests for valid and invalid datasets
	- Added a test for a simple motion for which kinematics can be easily derived.
	- Removed some hardcoded bits from the fixtures
- Expands kinematics tests to valid and invalid bboxes datasets
- Renames internal method for explicitness `_compute_approximate_derivative` ---> `_compute_approximate_time_derivative`
	- the reasoning being this makes it a bit more clear to developers and contributors how our wrapper to compute the approximate derivate and `xarray`'s method [differentiate](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.differentiate.html) are distinct. With `differentiate` you can compute the derivative along any single-coordinate dimension, not necessarily `time`.

**Question**
- Should we use [xarray's `differentiate`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.differentiate.html) in our kinematics wrappers? 

I am not sure if it existed or we were aware of it when we implemented these methods. If we decide to use it, we may need to change the tests, so that we continue to compute the expected kinematic variables with an 'independent' method. 

The simplest way would be to check the kinematic variables for a very simple motion case, for which the velocity, acceleration and displacement are very easy to infer / derive by hand. In a way this is similar to [the tests we currently have](https://github.com/neuroinformatics-unit/movement/blob/8c45183dd4f0beae7e9738287b084fafd8fff38c/tests/test_unit/test_kinematics.py#L19), except that we can more easily infer which values are expected. I added an example test like that called `test_kinematics_uniform_linear_motion` for reference, but right now it takes only the valid bboxes dataset.

## References

Overflow from #246 

## How has this PR been tested?

Tests pass locally and in CI.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

I updated the docstrings for the kinematics module so that the API reference is no longer specific to poses datasets.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

lochhh commented:
Cool! I wasn't aware of `xr.differentiate()`, which is much  simpler in code for our use case:
```python
result = data.differentiate("time")
```
vs.
```python
result = xr.apply_ufunc(
               np.gradient,
               data,
               data.coords["time"].values,
               kwargs={"axis": 0},
         )
result = result.reindex_like(data)
```
And in both cases, we no longer need to assume equidistant time-spacing.

lochhh commented on movement/analysis/kinematics.py:
```diff
@@ -69,16 +92,21 @@ def compute_acceleration(data: xr.DataArray) -> xr.DataArray:
     """Compute the derivative using numerical differentiation.
 
-    This assumes equidistant time spacing.
+    Uses numpy.gradient, which applies a second order accurate central
```
see above comment regarding intersphinx syntax

lochhh replied:
resolved in #269

lochhh commented on movement/analysis/kinematics.py:
```diff
@@ -69,16 +92,21 @@ def compute_acceleration(data: xr.DataArray) -> xr.DataArray:
 
+    See Also
+    --------
+    numpy.gradient: the underlying NumPy function used.
```
resolved in #269

sfmig commented:
thanks for the review @lochhh. A summary of the main changes below.

In the tests: 
- I added a fixture for a poses dataset with 2 individuals, each one with 3 keypoints, moving in a linear uniform motion. 
- The name of the added fixture is `valid_poses_dataset_uniform_linear_motion`, which doesn't really match the equivalent bboxes fixture but I didn't want to mess up existing tests that use `valid_poses_dataset` or `valid_bboxes_dataset`. I leave that to the refactoring fixtures PR (issue #222 ) if that is alright.
- I added a test for datasets with nans.

In the docstrings of the kinematics functions:
- I used monospace formatting for when we refer to arrays in a `movement` dataset.
- I removed `(x,y)` from cartesian, as the functions should work just fine if the trajectory is in 3D `(x,y,z)`.
- I replaced some references to "vector" by the term "array" where possible, else I clarified the relationship between them (e.g., a position array is made up of position vectors across time). Let me know if this is clear enough.
- I specified that the input position arrays should be in cartesian (even though this is not enforced in the implementation for now). 
- I added Notes to clarify what would happen when different position arrays are fed to each function. I am not sure if this is overkill / helpful, maybe we should remove it. I don't feel strongly either way.

Will re request review now

lochhh commented on tests/test_unit/test_kinematics.py:
```diff
@@ -1,112 +1,176 @@
+        ),
+    ],
+)
+def test_kinematics_uniform_linear_motion(
```
This test can be simplified as follows, so we don't need multiple asserts in a for loop:

```python
@pytest.mark.parametrize(
    "valid_dataset_uniform_linear_motion",
    [
        "valid_poses_dataset_uniform_linear_motion",
        "valid_bboxes_dataset",
    ],
)
@pytest.mark.parametrize(
    "kinematic_variable, expected_arrays",
    [
        (
            "displacement",
            [
                np.vstack([np.zeros((1, 2)), np.ones((9, 2))]),  # Individual 0
                np.multiply(
                    np.vstack([np.zeros((1, 2)), np.ones((9, 2))]),
                    np.array([1, -1]),
                ),  # Individual 1
            ],
        ),
        (
            "velocity",
            [
                np.ones((10, 2)),  # Individual 0
                np.multiply(
                    np.ones((10, 2)), np.array([1, -1])
                ),  # Individual 1
            ],
        ),
        (
            "acceleration",
            [
                np.zeros((10, 2)),  # Individual 0
                np.zeros((10, 2)),  # Individual 1
            ],
        ),
    ],
)
def test_kinematics_uniform_linear_motion(
    valid_dataset_uniform_linear_motion,
    kinematic_variable,
    expected_arrays,  # 2D: n_frames, n_space_dims
    request,
):
    """Test computed kinematics for a uniform linear motion case.

    Uniform linear motion means the individuals move along a line
    at constant velocity.

    We consider 2 individuals ("id_0" and "id_1"),
    tracked for 10 frames, along x and y:
    - id_0 moves along x=y line from the origin
    - id_1 moves along x=-y line from the origin
    - they both move one unit (pixel) along each axis in each frame

    If the dataset is a poses dataset, we consider 3 keypoints per individual
    (centroid, left, right), that are always in front of the centroid keypoint
    at 45deg from the trajectory.
    """
    position = request.getfixturevalue(
        valid_dataset_uniform_linear_motion
    ).position
    kinematic_variable = getattr(kinematics, f"compute_{kinematic_variable}")(
        position
    )
    expected_da = xr.DataArray(
        np.stack(expected_arrays, axis=1),  # Stack along the individuals axis
        dims=["time", "individuals", "space"],
    )
    if "keypoints" in position.coords:
        expected_da = expected_da.expand_dims(
            {"keypoints": position.coords["keypoints"].size}
        )
        expected_da = expected_da.transpose(
            "time", "individuals", "keypoints", "space"
        )
    # Compare the values of the kinematic_variable against the expected_da
    np.testing.assert_allclose(kinematic_variable.values, expected_da.values)
```

sfmig replied:
much cleaner and excellent point re the asserts in the loop, thank you!

lochhh commented on tests/test_unit/test_kinematics.py:
```diff
@@ -1,112 +1,176 @@
+    # compute kinematic array
+    valid_dataset = request.getfixturevalue(valid_dataset_with_nan)
+    position = valid_dataset.position
+    kinematic_array = getattr(kinematics, f"compute_{kinematic_variable}")(
```
```suggestion
    kinematic_variable = getattr(kinematics, f"compute_{kinematic_variable}")(
```
or change `kinematic_variable` in `test_kinematics_uniform_linear_motion` to `kinematic_array`.

lochhh commented on tests/test_unit/test_kinematics.py:
```diff
@@ -1,112 +1,176 @@
+        "valid_bboxes_dataset_with_nan",
+    ],
+)
+@pytest.mark.parametrize(
+    "kinematic_variable, expected_nans_per_individual",
+    [
+        ("displacement", {0: 5, 1: 0}),
+        ("velocity", {0: 6, 1: 0}),
+        ("acceleration", {0: 7, 1: 0}),
+    ],
+)
```
As above, I would use list instead of dict
```suggestion
@pytest.mark.parametrize(
    "kinematic_variable, expected_nans_per_individual",
    [
        ("displacement", [5, 0]),
        ("velocity", [6, 0]),
        ("acceleration", [7, 0]),
    ],
)
```

lochhh commented on tests/test_unit/test_kinematics.py:
```diff
@@ -1,112 +1,176 @@
+    kinematic_array = getattr(kinematics, f"compute_{kinematic_variable}")(
+        position
+    )
+
+    # compute n nans in kinematic array per individual
+    n_nans_kinematics_per_indiv = {
+        i: helpers.count_nans(kinematic_array.isel(individuals=i))
+        for i in range(valid_dataset.dims["individuals"])
+    }
+
+    # check number of nans per indiv is as expected in kinematic array
+    for i in range(valid_dataset.dims["individuals"]):
+        assert n_nans_kinematics_per_indiv[i] == (
+            expected_nans_per_individual[i]
+            * valid_dataset.dims["space"]
+            * valid_dataset.dims.get("keypoints", 1)
         )
```
```suggestion
    # compute n nans in kinematic array per individual
    n_nans_kinematics_per_indiv = [
        helpers.count_nans(kinematic_array.isel(individuals=i))
        for i in range(valid_dataset.dims["individuals"])
    ]
    # expected nans per individual adjusted for space and keypoints dimensions
    expected_nans_adjusted = [
        n
        * valid_dataset.dims["space"]
        * valid_dataset.dims.get("keypoints", 1)
        for n in expected_nans_per_individual
    ]
    # check number of nans per individual is as expected in kinematic array
    np.testing.assert_array_equal(
        n_nans_kinematics_per_indiv, expected_nans_adjusted
    )
```
And we can assert the array as a whole

sfmig replied:
looks so clear, thanks! ✨ 🤩

lochhh reviewed (approved):
Thanks so much for this @sfmig ! ❤️ the restructured tests! I'm happy to have the additional notes in the Notes section - I guess people can read or ignore. Mostly minor comments and suggestions, feel free to take or leave. I also tried to avoid multiple assertions in a for loop where possible, as this makes the code inefficient to debug - the test will fail and terminate on the first failure and we'd need to re-run the test to uncover issues one-by-one.

sfmig commented on movement/analysis/kinematics.py:
```diff
@@ -33,22 +49,34 @@ def compute_displacement(data: xr.DataArray) -> xr.DataArray:
-    """Compute the velocity in cartesian ``(x,y)`` coordinates.
+    """Compute velocity array in cartesian coordinates.
+
+    The velocity array is the first time derivative of the position
```
Oh good point. I was thinking about this sentence as a general statement, "time" referring to the physical quantity, rather than the specific name of the array dimension. But it is true that we have changed the sentence to say "array"...

I am inclined to leave it without formatting, in line with the kinematic variable just before it in the same sentence (velocity, acceleration). I changed it to "time-derivative" to be a bit more explicit that it the general term we are talking about.

sfmig commented on movement/analysis/kinematics.py:
```diff
@@ -87,24 +128,26 @@ def compute_acceleration(data: xr.DataArray) -> xr.DataArray:
     -------
     xarray.DataArray
-        An xarray DataArray containing the derived variable.
+        An xarray DataArray containing the time derivative of the
```
(see first answer to this point above)
